src/pymor/algorithms/lrradi.py

```diff
@@ -0,0 +1,343 @@
+            Z.append(V)
+            VB = V.dot(B)
+            Yt = np.identity(len(C)) - 1 / (2 * shifts[j_shift].real) * (VB @ VB.T)
+            Y = spla.block_diag(Y, Yt)
+            if not trans:
+                EVYt = E.apply(V).lincomb(np.linalg.inv(Yt))
+            else:
+                EVYt = E.apply_adjoint(V).lincomb(np.linalg.inv(Yt))
+            RF += EVYt * np.sqrt(-2*shifts[j_shift].real)
+            K += EVYt.lincomb(VB.T)
+            j += 1
+        else:
+            Z.append(V.real)
+            Z.append(V.imag)
+            Vr = V.real.dot(B)
+            Vi = V.imag.dot(B)
+            sa = np.abs(shifts[j_shift])
+            F1 = np.block([
+                [-shifts[j_shift].real/sa * Vr - shifts[j_shift].imag/sa * Vi],
+                [shifts[j_shift].imag/sa * Vr - shifts[j_shift].real/sa * Vi]
+            ])
+            F2 = np.block([
+                [Vr],
+                [Vi]
+            ])
+            F3 = np.block([
+                [shifts[j_shift].imag/sa * np.identity(len(C))],
+                [shifts[j_shift].real/sa * np.identity(len(C))]
+            ])
+            Yt = spla.block_diag(np.identity(len(C)), 0.5 * np.identity(len(C))) \
+                - 1 / (4 * shifts[j_shift].real) * (F1 @ F1.T) \
+                - 1 / (4 * shifts[j_shift].real) * (F2 @ F2.T) \
+                - 1 / 2 * (F3 @ F3.T)
+            Y = spla.block_diag(Y, Yt)
+            EVYt = E.apply(cat_arrays([V.real, V.imag])).lincomb(np.linalg.inv(Yt))
+            RF += EVYt[:len(C)] * np.sqrt(-2 * shifts[j_shift].real)
+            K += EVYt.lincomb(F2.T)
+            j += 2
+        j_shift += 1
+        res = np.linalg.norm(RF.gramian(), ord=2)
+        logger.info(f'Relative residual at step {j}: {res/init_res:.5e}')
+        if j_shift >= shifts.size:
+            shifts = iteration_shifts(A, E, B, RF, K, Z, shift_options)
+            j_shift = 0
+    # transform solution to lrcf
+    cf = spla.cholesky(Y)
+    Z_cf = Z.lincomb(spla.solve_triangular(cf, np.identity(len(Z))).T)
+    return Z_cf
+
+
+def hamiltonian_shifts_init(A, E, B, C, shift_options):
+    """Find starting shift parameters for low-rank RADI iteration using
+    Galerkin projection on spaces spanned by LR-RADI iterates.
+
+    See [BBKS18]_, pp. 318-321.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A from the corresponding Riccati equation.
+    E
+        The |Operator| E from the corresponding Riccati equation.
+    B
+        The |VectorArray| B from the corresponding Riccati equation.
+    C
+        The |VectorArray| C from the corresponding Riccati equation.
+    shift_options
+        The shift options to use (see :func:`ricc_lrcf_solver_options`).
+
+    Returns
+    -------
+    shifts
+        A |NumPy array| containing a set of stable shift parameters.
+    """
+    random_state = get_random_state(seed=shift_options['init_seed'])
+    for j in range(shift_options['init_maxiter']):
```

Comment:
If you want, you can use an underscore instead of `j` (i.e., `for _ in ...`) to make it clear that the iteration index is not used